File: C++/Problems/Problems/problem_1331.cpp
#include "problem_1331.h"

#include <unordered_map>

bool problem_1331::test() {
    std::vector arr { 40, 10, 20, 30 };

    std::vector expected { 4, 1, 2, 3 };

    auto result = arrayRankTransform(arr);

    print1DVector(result);

    return expected == result;
}

std::vector<int> problem_1331::arrayRankTransform(std::vector<int> &arr) {
    int n = arr.size();
    std::unordered_map<int, int> dic;
    auto sorted = arr;
    std::sort(sorted.begin(), sorted.end());
    int cur = 1;
    for (int i = 0; i < n; ++i)
    {
        if (!dic.contains(sorted[i]))
        {
            dic[sorted[i]] = cur;
            cur += 1;
        }
    }
    std::vector<int> result(n);
    for (int i = 0; i < n; ++i)
        result[i] = dic[arr[i]];

    return result;
}

File: Python/Problems/Problems/problem1331.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem


class Problem1331(Problem):
    def test(self) -> bool:
        arr = [40, 10, 20, 30]

        expected = [4, 1, 2, 3]

        result = self.arrayRankTransform(arr)

        print1DArray(result)

        return expected == result

    def arrayRankTransform(self, arr: List[int]) -> List[int]:
        n = len(arr)
        dic = dict()
        sorted = list(arr)
        sorted.sort()
        cur = 1
        for i in range(n):
            if sorted[i] not in dic:
                dic[sorted[i]] = cur
                cur += 1
        result = [0] * n
        for i in range(n):
            result[i] = dic[arr[i]]

        return result